airbyte-workers/src/main/java/io/airbyte/workers/DefaultDiscoverCatalogWorker.java

```diff
@@ -43,49 +49,60 @@
   private static final Logger LOGGER = LoggerFactory.getLogger(DefaultDiscoverCatalogWorker.class);
 
   private final IntegrationLauncher integrationLauncher;
+  private final AirbyteStreamFactory streamFactory;
 
   private volatile Process process;
 
-  public DefaultDiscoverCatalogWorker(final IntegrationLauncher integrationLauncher) {
+  public DefaultDiscoverCatalogWorker(final IntegrationLauncher integrationLauncher, final AirbyteStreamFactory streamFactory) {
     this.integrationLauncher = integrationLauncher;
+    this.streamFactory = streamFactory;
+  }
+
+  public DefaultDiscoverCatalogWorker(final IntegrationLauncher integrationLauncher) {
+    this(integrationLauncher, new DefaultAirbyteStreamFactory());
   }
 
   @Override
   public OutputAndStatus<StandardDiscoverCatalogOutput> run(final StandardDiscoverCatalogInput discoverSchemaInput,
                                                             final Path jobRoot) {
-    try {
-      return runInternal(discoverSchemaInput, jobRoot);
-    } catch (final Exception e) {
-      LOGGER.error("Error while discovering schema", e);
-      return new OutputAndStatus<>(JobStatus.FAILED);
-    }
-  }
 
-  private OutputAndStatus<StandardDiscoverCatalogOutput> runInternal(final StandardDiscoverCatalogInput discoverSchemaInput,
-                                                                     final Path jobRoot)
-      throws IOException, WorkerException {
     final JsonNode configDotJson = discoverSchemaInput.getConnectionConfiguration();
-
     IOs.writeFile(jobRoot, WorkerConstants.TAP_CONFIG_JSON_FILENAME, Jsons.serialize(configDotJson));
 
-    process = integrationLauncher.discover(jobRoot, WorkerConstants.TAP_CONFIG_JSON_FILENAME)
-        // TODO: we shouldn't trust the tap does not pollute stdout
-        .redirectOutput(jobRoot.resolve(WorkerConstants.CATALOG_JSON_FILENAME).toFile())
-        .start();
-
-    LineGobbler.gobble(process.getErrorStream(), LOGGER::error);
-
-    WorkerUtils.gentleClose(process, 1, TimeUnit.MINUTES);
-
-    int exitCode = process.exitValue();
-
-    if (exitCode == 0) {
-      return new OutputAndStatus<>(
-          JobStatus.SUCCEEDED,
-          new StandardDiscoverCatalogOutput()
-              .withCatalog(readCatalog(jobRoot)));
-    } else {
-      LOGGER.debug("Discovery job subprocess finished with exit code {}", exitCode);
+    try {
```

Comment:
all of the workers (and their tests) are beginning to look kinda the same interally. i think there's room to DRY them.